web/php/Modules/Account/PM/PMComposeModule.php

```diff
@@ -1,61 +1,77 @@
 <?php
 
-namespace Wikidot\Modules\Account\PM;
-
-
+declare(strict_types=1);
 
+namespace Wikidot\Modules\Account\PM;
 
-use Wikidot\DB\PrivateMessagePeer;
 
 use Wikidot\Utils\AccountBaseModule;
 use Wikidot\Utils\ProcessException;
 use Wikijump\Models\User;
+use Wikijump\Models\UserMessage;
 
+/**
+ * AJAX view for message composition.
+ * @package Wikidot\Modules\Account\PM
+ */
 class PMComposeModule extends AccountBaseModule
 {
 
+    /**
+     * Write or resume writing a private message.
+     * @param $runData
+     * @throws ProcessException
+     */
     public function build($runData)
     {
-
-        $user = $runData->getUser();
+        $user = $runData->user();
 
         $pl = $runData->getParameterList();
-        $replyMessageId = $pl->getParameterValue("replyMessageId", "AMODULE");
+        $replyMessageId = $pl->getParameterValue('replyMessageId', 'AMODULE');
 
-        $continueMessageId = $pl->getParameterValue("continueMessageId", "AMODULE");
-        $toUserId = $pl->getParameterValue("toUserId");
+        $continueMessageId = $pl->getParameterValue('continueMessageId', 'AMODULE');
+        $toUserId = $pl->getParameterValue('toUserId');
 
-        if ($replyMessageId) {
-            $message = PrivateMessagePeer::instance()->selectByPrimaryKey($replyMessageId);
+        if ($replyMessageId)
+        {
+            $message = UserMessage::find($replyMessageId);
 
-            if ($message == null || $message->getToUserId() != $user->id) {
-                throw new ProcessException(_("Error getting orginal message."), "no_reply_message");
+            if ($message == null || $message->recipient->id != $user->id)
+            {
+                throw new ProcessException(_('Error getting orginal message.'), 'no_reply_message');
             }
-            $runData->ajaxResponseAdd("toUserId", $message->getFromUserId());
-            $runData->ajaxResponseAdd("toUserName", $message->getFromUser()->username);
-            $subject = $message->getSubject();
-            $subject = preg_replace("/^Re: /", '', $subject);
-            $runData->contextAdd("subject", "Re: ".$subject);
-        } elseif ($continueMessageId) {
-            $message = PrivateMessagePeer::instance()->selectByPrimaryKey($continueMessageId);
-
-            if ($message == null || $message->getFromUserId() != $user->id()) {
-                throw new ProcessException(_("Error getting orginal message."), "no_reply_message");
+
+            $subject = $message->subject;
+            $subject = preg_replace('/^Re: /', '', $subject);
```

Comment:
I think this would be cleaner rather than a substitute then concat:
```php
if (!str_starts_with($subject, 'Re: ')) {
    $subject = 'Re: ' . $subject;
}
```

Reply:
See, that would make sense, but that's not what this code segment does. It's basically "ensuring" that there's only one `Re: ` prepended to a message: It removes the original in line 45 and then concatenates "Re: " to the subject when it's done in line 49.

You might say "That's dumb, why are we removing the string and then adding it back again?" and I would say "I don't know why we're doing that."

Reply:
Right, it ensures there's only one `Re: `, but we can do the same by only prepending it if it doesn't already have `Re: ` at the start.

Reply:
This is changed in the next commit.